packages/fx-core/src/plugins/solution/fx-solution/arm.ts

```diff
@@ -56,93 +61,39 @@ const parameterFileNameTemplateNew = "azure.parameters.@envName.json";
-      ? path.join(ctx.root, templateFolderNew)
-      : path.join(ctx.root, baseFolder, templateFolder);
-    await fs.ensureDir(templateFolderPath);
-    await fs.writeFile(
-      path.join(templateFolderPath, bicepOrchestrationFileName),
-      bicepOrchestrationFileContent
-    );
-
-    // Output bicep module files from each resource plugin
-    const modulesFolderPath = isMultiEnvEnabled()
-      ? path.join(templateFolderPath, modulesFolder)
-      : templateFolderPath;
-    await fs.ensureDir(modulesFolderPath);
-    for (const module of moduleFiles) {
-      // module[0] contains relative path to template folder, e.g. "./modules/frontendHosting.bicep"
-      await fs.writeFile(path.join(templateFolderPath, module[0]), module[1]);
+  let result: Result<void, FxError>;
+  ctx.telemetryReporter?.sendTelemetryEvent(SolutionTelemetryEvent.GenerateArmTemplateStart, {
+    [SolutionTelemetryProperty.Component]: SolutionTelemetryComponentName,
+  });
+  try {
+    result = await doGenerateArmTemplate(ctx);
+    if (result.isOk()) {
+      ctx.telemetryReporter?.sendTelemetryEvent(SolutionTelemetryEvent.GenerateArmTemplate, {
+        [SolutionTelemetryProperty.Component]: SolutionTelemetryComponentName,
+        [SolutionTelemetryProperty.Success]: SolutionTelemetrySuccess.Yes,
+      });
+    } else {
+      sendErrorTelemetryThenReturnError(
+        SolutionTelemetryEvent.GenerateArmTemplate,
+        result.error,
+        ctx.telemetryReporter
+      );
     }
-
-    // Output parameter file
-    const parameterTemplateFolderPath = isMultiEnvEnabled()
-      ? path.join(ctx.root, templateFolderNew)
-      : path.join(ctx.root, baseFolder, parameterFolder);
-    const parameterTemplateFilePath = path.join(
-      parameterTemplateFolderPath,
-      parameterTemplateFileName
+  } catch (error) {
+    result = err(
+      returnSystemError(
+        error,
+        PluginDisplayName.Solution,
+        SolutionError.FailedToDeployArmTemplatesToAzure
+      )
     );
-    const parameterFileContent = bicepOrchestrationTemplate.getParameterFileContent();
-    await fs.ensureDir(parameterTemplateFolderPath);
-    await fs.writeFile(parameterTemplateFilePath, parameterFileContent);
-
-    // Output .gitignore file
-    const gitignoreContent = await fs.readFile(
-      path.join(getTemplatesFolder(), "plugins", "solution", "armGitignore"),
-      ConstantString.UTF8Encoding
+    sendErrorTelemetryThenReturnError(
+      SolutionTelemetryEvent.GenerateArmTemplate,
+      result.error,
+      ctx.telemetryReporter
     );
-    const gitignoreFileName = ".gitignore";
-    const gitignoreFilePath = isMultiEnvEnabled()
-      ? path.join(ctx.root, templateFolderNew, gitignoreFileName)
-      : path.join(ctx.root, baseFolder, gitignoreFileName);
-    if (!(await fs.pathExists(gitignoreFilePath))) {
-      await fs.writeFile(gitignoreFilePath, gitignoreContent);
-    }
   }
-
-  return ok(undefined); // Nothing to return when success
+  return result;
```

Comment:
You can return sendErrorTelemetryThenReturnError(...) instead.

Reply:
offline synced, the result is shared in between exception handling logic and regular logic, so will keep existing implementation for now.